runtime/protocol/http/common/src/aws/smithy/kotlin/runtime/http/middleware/Retry.kt

```diff
@@ -11,31 +11,67 @@ import aws.smithy.kotlin.runtime.http.HttpBody
 import aws.smithy.kotlin.runtime.http.HttpClientFeatureFactory
 import aws.smithy.kotlin.runtime.http.operation.SdkHttpOperation
 import aws.smithy.kotlin.runtime.http.operation.deepCopy
+import aws.smithy.kotlin.runtime.http.operation.getLogger
 import aws.smithy.kotlin.runtime.http.request.HttpRequestBuilder
+import aws.smithy.kotlin.runtime.logging.Logger
+import aws.smithy.kotlin.runtime.retries.RetryDirective
 import aws.smithy.kotlin.runtime.retries.RetryPolicy
 import aws.smithy.kotlin.runtime.retries.RetryStrategy
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * Retry requests with the given strategy and policy
+ * @param strategy the [RetryStrategy] to retry failed requests with
+ * @param policy the [RetryPolicy] used to determine when to retry
+ */
+@InternalApi
+class Retry(
+    private val strategy: RetryStrategy,
+    private val policy: RetryPolicy<Any?>
+) : Feature {
 
-class RetryFeature(private val strategy: RetryStrategy, private val policy: RetryPolicy<Any?>) : Feature {
     class Config {
         var strategy: RetryStrategy? = null
         var policy: RetryPolicy<Any?>? = null
     }
 
-    companion object Feature : HttpClientFeatureFactory<Config, RetryFeature> {
-        override val key: FeatureKey<RetryFeature> = FeatureKey("RetryFeature")
+    companion object Feature : HttpClientFeatureFactory<Config, Retry> {
+        override val key: FeatureKey<Retry> = FeatureKey("Retry")
 
-        override fun create(block: Config.() -> Unit): RetryFeature {
+        override fun create(block: Config.() -> Unit): Retry {
             val config = Config().apply(block)
             val strategy = requireNotNull(config.strategy) { "strategy is required" }
             val policy = requireNotNull(config.policy) { "policy is required" }
-            return RetryFeature(strategy, policy)
+            return Retry(strategy, policy)
+        }
+    }
+
+    /**
+     * Wrapper around [policy] that logs termination decisions
+     */
+    private class PolicyLogger(
+        private val policy: RetryPolicy<Any?>,
+        private val logger: Logger,
+    ) : RetryPolicy<Any?> {
+        override fun evaluate(result: Result<Any?>): RetryDirective = policy.evaluate(result).also {
+            if (it is RetryDirective.TerminateAndFail) {
+                logger.debug { "request failed with non-retryable error" }
+            }
         }
     }
 
     override fun <I, O> install(operation: SdkHttpOperation<I, O>) {
         operation.execution.finalize.intercept { req, next ->
             if (req.subject.isRetryable) {
-                strategy.retry(policy) {
+                var attempt = 1
+                val logger = req.context.getLogger("Retry")
+                val wrappedPolicy = PolicyLogger(policy, logger)
+
+                strategy.retry(wrappedPolicy) {
+                    if (attempt > 1) {
+                        logger.debug { "retrying request, attempt $attempt" }
+                    }
```

Comment:
**Comment**: This feels weird. We're duplicating the attempt-counting machinery that already exists inside the retry strategy. Is this because request context isn't available outside of the `intercept` method? And we need request context to get a context-aware logger?

If so, then I wonder if request context could actually be available from `CoroutineContext` or something more global that would enable easier access to things that aren't `intercept` methods.

Logically, it seems like this logging ("retrying request, attempt _n_") should happen regardless of whether we're in a request or not. It feels like it belongs inside the strategy, not the feature.

Reply:
FYI holding off on this PR as it seems the work is still in progress, LMK if this is not the case.

Reply:
@kggilmer No it's ready, this is a point of discussion for sure but otherwise it's done

Reply:
Circling back, I can see some pros in moving logging to the middleware and some other pros in keeping it part of the strategies. I think the difference comes down to style so I won't block on this point.